MCP: in a codespace, filepaths are rewritten for windows.

<!-- ⚠️⚠️ Do Not Delete This! bug_report_template ⚠️⚠️ -->
<!-- Please read our Rules of Conduct: https://opensource.microsoft.com/codeofconduct/ -->
<!-- 🕮 Read our guide about submitting issues: https://github.com/microsoft/vscode/wiki/Submitting-Bugs-and-Suggestions -->
<!-- 🔎 Search existing issues to avoid creating duplicates. -->
<!-- 🧪 Test using the latest Insiders build to see if your issue has already been fixed: https://code.visualstudio.com/insiders/ -->
<!-- 💡 Instead of creating your report here, use 'Report Issue' from the 'Help' menu in VS Code to pre-fill useful information. -->
<!-- 🔧 Launch with `code --disable-extensions` to check. -->
Does this issue occur when all extensions are disabled?: 

<!-- 🪓 If you answered No above, use 'Help: Start Extension Bisect' from Command Palette to try to identify the cause. -->
<!-- 📣 Issues caused by an extension need to be reported directly to the extension publisher. The 'Help > Report Issue' dialog can assist with this. -->
- VS Code Version:  1.99-insiders
- OS Version: codespace/windows 11

Steps to Reproduce:

1. create a dummy mcp javascript file and place it in the file system. In this case "../cli/built/genaiscript.cjs".
2. update `.vscode/mcp.json` to mount a MCP server on that file.
```
{
    "servers": {
        "genaiscript": {
            "type": "stdio",
            "command": "node",
            "args": ["${workspaceFolder}/../cli/built/genaiscript.cjs", "mcp"],
            "envFile": "${workspaceFolder}/.env"
        }
    }
}
```

3. Try to launch the server. If your file is loaded and contains a mcp server, it should succeed.


On WIndows, the filepath is resolved correctly and the MCP server starts correctly.
In the codespace, the filepath somewhat becomes a windows path and fails to load.

```
2025-03-20 14:41:00.910 [info] Error: Cannot find module 'C:\workspaces\genaiscript\packages\cli\built\genaiscript.cjs'
```

--- a/src/vs/platform/log/common/log.ts
+++ b/src/vs/platform/log/common/log.ts
@@ -57,6 +57,10 @@ export interface ILogger extends IDisposable {
 	flush(): void;
 }
 
+export function canLog(loggerLevel: LogLevel, messageLevel: LogLevel): boolean {
+	return loggerLevel !== LogLevel.Off && loggerLevel <= messageLevel;
+}
+
 export function log(logger: ILogger, level: LogLevel, message: string): void {
 	switch (level) {
 		case LogLevel.Trace: logger.trace(message); break;
@@ -267,7 +271,7 @@ export abstract class AbstractLogger extends Disposable implements ILogger {
 	}
 
 	protected checkLogLevel(level: LogLevel): boolean {
-		return this.level !== LogLevel.Off && this.level <= level;
+		return canLog(this.level, level);
 	}
 
 	protected canLog(level: LogLevel): boolean {

--- a/src/vs/workbench/api/browser/mainThreadMcp.ts
+++ b/src/vs/workbench/api/browser/mainThreadMcp.ts
@@ -3,13 +3,15 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
+import { disposableTimeout } from '../../../base/common/async.js';
 import { Emitter } from '../../../base/common/event.js';
 import { Disposable, DisposableMap } from '../../../base/common/lifecycle.js';
 import { ISettableObservable, observableValue } from '../../../base/common/observable.js';
+import { LogLevel } from '../../../platform/log/common/log.js';
 import { IMcpMessageTransport, IMcpRegistry } from '../../contrib/mcp/common/mcpRegistryTypes.js';
 import { McpCollectionDefinition, McpConnectionState, McpServerDefinition, McpServerTransportType } from '../../contrib/mcp/common/mcpTypes.js';
 import { MCP } from '../../contrib/mcp/common/modelContextProtocol.js';
-import { ExtensionHostKind } from '../../services/extensions/common/extensionHostKind.js';
+import { ExtensionHostKind, extensionHostKindToString } from '../../services/extensions/common/extensionHostKind.js';
 import { IExtHostContext, extHostNamedCustomer } from '../../services/extensions/common/extHostCustomers.js';
 import { ExtHostContext, MainContext, MainThreadMcpShape } from '../common/extHost.protocol.js';
 
@@ -48,6 +50,7 @@ export class MainThreadMcp extends Disposable implements MainThreadMcpShape {
 			start: (collection, _serverDefiniton, resolveLaunch) => {
 				const id = ++this._serverIdCounter;
 				const launch = new ExtHostMcpServerLaunch(
+					_extHostContext.extensionHostKind,
 					() => proxy.$stopMcp(id),
 					msg => proxy.$sendMessage(id, JSON.stringify(msg)),
 				);
@@ -93,8 +96,13 @@ export class MainThreadMcp extends Disposable implements MainThreadMcpShape {
 		}
 	}
 
-	$onDidPublishLog(id: number, log: string): void {
-		this._servers.get(id)?.pushLog(log);
+	$onDidPublishLog(id: number, level: LogLevel, log: string): void {
+		if (typeof level === 'string') {
+			level = LogLevel.Info;
+			log = level as unknown as string;
+		}
+
+		this._servers.get(id)?.pushLog(level, log);
 	}
 
 	$onDidReceiveMessage(id: number, message: string): void {
@@ -106,14 +114,14 @@ export class MainThreadMcp extends Disposable implements MainThreadMcpShape {
 class ExtHostMcpServerLaunch extends Disposable implements IMcpMessageTransport {
 	public readonly state = observableValue<McpConnectionState>('mcpServerState', { state: McpConnectionState.Kind.Starting });
 
-	private readonly _onDidLog = this._register(new Emitter<string>());
+	private readonly _onDidLog = this._register(new Emitter<{ level: LogLevel; message: string }>());
 	public readonly onDidLog = this._onDidLog.event;
 
 	private readonly _onDidReceiveMessage = this._register(new Emitter<MCP.JSONRPCMessage>());
 	public readonly onDidReceiveMessage = this._onDidReceiveMessage.event;
 
-	pushLog(log: string): void {
-		this._onDidLog.fire(log);
+	pushLog(level: LogLevel, message: string): void {
+		this._onDidLog.fire({ message, level });
 	}
 
 	pushMessage(message: string): void {
@@ -121,7 +129,7 @@ class ExtHostMcpServerLaunch extends Disposable implements IMcpMessageTransport 
 		try {
 			parsed = JSON.parse(message);
 		} catch (e) {
-			this.pushLog(`Failed to parse message: ${JSON.stringify(message)}`);
+			this.pushLog(LogLevel.Warning, `Failed to parse message: ${JSON.stringify(message)}`);
 		}
 
 		if (parsed) {
@@ -130,15 +138,20 @@ class ExtHostMcpServerLaunch extends Disposable implements IMcpMessageTransport 
 	}
 
 	constructor(
+		extHostKind: ExtensionHostKind,
 		public readonly stop: () => void,
 		public readonly send: (message: MCP.JSONRPCMessage) => void,
 	) {
 		super();
+
+		this._register(disposableTimeout(() => {
+			this.pushLog(LogLevel.Info, `Starting server from ${extensionHostKindToString(extHostKind)}extension host`);
+		}));
 	}
 
 	public override dispose(): void {
 		if (McpConnectionState.isRunning(this.state.get())) {
-			this.pushLog('Extension host shut down, server will stop.');
+			this.pushLog(LogLevel.Warning, 'Extension host shut down, server will stop.');
 			this.state.set({ state: McpConnectionState.Kind.Stopped }, undefined);
 		}
 

--- a/src/vs/workbench/api/browser/mainThreadMcp.ts
+++ b/src/vs/workbench/api/browser/mainThreadMcp.ts
@@ -145,7 +145,7 @@ class ExtHostMcpServerLaunch extends Disposable implements IMcpMessageTransport 
 		super();
 
 		this._register(disposableTimeout(() => {
-			this.pushLog(LogLevel.Info, `Starting server from ${extensionHostKindToString(extHostKind)}extension host`);
+			this.pushLog(LogLevel.Info, `Starting server from ${extensionHostKindToString(extHostKind)} extension host`);
 		}));
 	}
 

--- a/src/vs/workbench/api/common/extHost.protocol.ts
+++ b/src/vs/workbench/api/common/extHost.protocol.ts
@@ -2980,7 +2980,7 @@ export interface ExtHostMcpShape {
 
 export interface MainThreadMcpShape {
 	$onDidChangeState(id: number, state: McpConnectionState): void;
-	$onDidPublishLog(id: number, log: string): void;
+	$onDidPublishLog(id: number, level: LogLevel, log: string): void;
 	$onDidReceiveMessage(id: number, message: string): void;
 	$upsertMcpCollection(collection: McpCollectionDefinition.FromExtHost, servers: Dto<McpServerDefinition>[]): void;
 	$deleteMcpCollection(collectionId: string): void;

--- a/src/vs/workbench/api/common/extHostMcp.ts
+++ b/src/vs/workbench/api/common/extHostMcp.ts
@@ -16,6 +16,7 @@ import { StorageScope } from '../../../platform/storage/common/storage.js';
 import { extensionPrefixedIdentifier, McpCollectionDefinition, McpConnectionState, McpServerDefinition, McpServerLaunch, McpServerTransportSSE, McpServerTransportType } from '../../contrib/mcp/common/mcpTypes.js';
 import { ExtHostMcpShape, MainContext, MainThreadMcpShape } from './extHost.protocol.js';
 import { IExtHostRpcService } from './extHostRpcService.js';
+import { LogLevel } from '../../../platform/log/common/log.js';
 
 export const IExtHostMpcService = createDecorator<IExtHostMpcService>('IExtHostMpcService');
 
@@ -224,7 +225,7 @@ class McpSSEHandle extends Disposable {
 			});
 
 			if (res.status >= 300) {
-				this._proxy.$onDidPublishLog(this._id, `${res.status} status sending message to ${this._postEndpoint}: ${await this._getErrText(res)}`);
+				this._proxy.$onDidPublishLog(this._id, LogLevel.Warning, `${res.status} status sending message to ${this._postEndpoint}: ${await this._getErrText(res)}`);
 			}
 		} catch (err) {
 			// ignored

--- a/src/vs/workbench/api/node/extHostMpcNode.ts
+++ b/src/vs/workbench/api/node/extHostMpcNode.ts
@@ -6,10 +6,10 @@
 import { ChildProcessWithoutNullStreams, spawn } from 'child_process';
 import { readFile } from 'fs/promises';
 import { homedir } from 'os';
-import { PassThrough } from 'stream';
 import { parseEnvFile } from '../../../base/common/envfile.js';
 import { URI } from '../../../base/common/uri.js';
 import { StreamSplitter } from '../../../base/node/nodeStreams.js';
+import { LogLevel } from '../../../platform/log/common/log.js';
 import { McpConnectionState, McpServerLaunch, McpServerTransportStdio, McpServerTransportType } from '../../contrib/mcp/common/mcpTypes.js';
 import { ExtHostMcpService } from '../common/extHostMcp.js';
 import { IExtHostRpcService } from '../common/extHostRpcService.js';
@@ -47,7 +47,6 @@ export class NodeExtHostMpcService extends ExtHostMcpService {
 	override $sendMessage(id: number, message: string): void {
 		const nodeServer = this.nodeServers.get(id);
 		if (nodeServer) {
-			this._proxy.$onDidPublishLog(id, '[Client Says] ' + message.toString());
 			nodeServer.child.stdin.write(message + '\n');
 		} else {
 			super.$sendMessage(id, message);
@@ -94,19 +93,14 @@ export class NodeExtHostMpcService extends ExtHostMcpService {
 
 		this._proxy.$onDidChangeState(id, { state: McpConnectionState.Kind.Starting });
 
-		const debug = new PassThrough();
-		debug.on('data', line => {
-			this._proxy.$onDidPublishLog(id, '[Server Says] ' + line.toString());
-		});
-
-		child.stdout.pipe(new StreamSplitter('\n')).pipe(debug).on('data', line => this._proxy.$onDidReceiveMessage(id, line.toString()));
+		child.stdout.pipe(new StreamSplitter('\n')).on('data', line => this._proxy.$onDidReceiveMessage(id, line.toString()));
 
 		child.stdin.on('error', onError);
 		child.stdout.on('error', onError);
 
 		// Stderr handling is not currently specified https://github.com/modelcontextprotocol/specification/issues/177
 		// Just treat it as generic log data for now
-		child.stderr.pipe(new StreamSplitter('\n')).on('data', line => this._proxy.$onDidPublishLog(id, line.toString()));
+		child.stderr.pipe(new StreamSplitter('\n')).on('data', line => this._proxy.$onDidPublishLog(id, LogLevel.Warning, `[server stderr] ${line}`));
 
 		child.on('spawn', () => this._proxy.$onDidChangeState(id, { state: McpConnectionState.Kind.Running }));
 

--- a/src/vs/workbench/contrib/mcp/common/mcpConfigPathsService.ts
+++ b/src/vs/workbench/contrib/mcp/common/mcpConfigPathsService.ts
@@ -65,7 +65,7 @@ export class McpConfigPathsService extends Disposable implements IMcpConfigPaths
 		@IWorkspaceContextService workspaceContextService: IWorkspaceContextService,
 		@IProductService productService: IProductService,
 		@ILabelService labelService: ILabelService,
-		@IWorkbenchEnvironmentService environmentService: IWorkbenchEnvironmentService,
+		@IWorkbenchEnvironmentService private readonly _environmentService: IWorkbenchEnvironmentService,
 		@IRemoteAgentService remoteAgentService: IRemoteAgentService,
 		@IPreferencesService preferencesService: IPreferencesService,
 	) {
@@ -90,6 +90,7 @@ export class McpConfigPathsService extends Disposable implements IMcpConfigPaths
 				label: localize('mcp.configuration.workspaceValue', 'From your workspace'),
 				scope: StorageScope.WORKSPACE,
 				order: McpCollectionSortOrder.Workspace,
+				remoteAuthority: _environmentService.remoteAuthority,
 				uri: workspaceConfig,
 				section: ['settings', mcpConfigurationSection],
 			},
@@ -101,7 +102,7 @@ export class McpConfigPathsService extends Disposable implements IMcpConfigPaths
 		this._paths = observableValue('mcpConfigPaths', initialPaths.filter(isDefined));
 
 		remoteAgentService.getEnvironment().then((env) => {
-			const label = environmentService.remoteAuthority ? labelService.getHostLabel(Schemas.vscodeRemote, environmentService.remoteAuthority) : 'Remote';
+			const label = _environmentService.remoteAuthority ? labelService.getHostLabel(Schemas.vscodeRemote, _environmentService.remoteAuthority) : 'Remote';
 
 			this._paths.set([
 				...this.paths.get(),
@@ -113,7 +114,7 @@ export class McpConfigPathsService extends Disposable implements IMcpConfigPaths
 					scope: StorageScope.PROFILE,
 					order: McpCollectionSortOrder.User + McpCollectionSortOrder.RemoteBoost,
 					uri: env?.settingsPath,
-					remoteAuthority: environmentService.remoteAuthority,
+					remoteAuthority: _environmentService.remoteAuthority,
 					section: [mcpConfigurationSection],
 				}
 			], undefined);
@@ -141,6 +142,7 @@ export class McpConfigPathsService extends Disposable implements IMcpConfigPaths
 			target: ConfigurationTarget.WORKSPACE_FOLDER,
 			label: workspaceFolder.name,
 			scope: StorageScope.WORKSPACE,
+			remoteAuthority: this._environmentService.remoteAuthority,
 			order: McpCollectionSortOrder.WorkspaceFolder,
 			uri: URI.joinPath(workspaceFolder.uri, FOLDER_SETTINGS_PATH, '../mcp.json'),
 			workspaceFolder,

--- a/src/vs/workbench/contrib/mcp/common/mcpRegistryTypes.ts
+++ b/src/vs/workbench/contrib/mcp/common/mcpRegistryTypes.ts
@@ -8,7 +8,7 @@ import { IDisposable } from '../../../../base/common/lifecycle.js';
 import { IObservable } from '../../../../base/common/observable.js';
 import { ConfigurationTarget } from '../../../../platform/configuration/common/configuration.js';
 import { createDecorator } from '../../../../platform/instantiation/common/instantiation.js';
-import { ILogger } from '../../../../platform/log/common/log.js';
+import { ILogger, LogLevel } from '../../../../platform/log/common/log.js';
 import { StorageScope } from '../../../../platform/storage/common/storage.js';
 import { IWorkspaceFolderData } from '../../../../platform/workspace/common/workspace.js';
 import { IResolvedValue } from '../../../services/configurationResolver/common/configurationResolverExpression.js';
@@ -20,7 +20,7 @@ export const IMcpRegistry = createDecorator<IMcpRegistry>('mcpRegistry');
 /** Message transport to a single MCP server. */
 export interface IMcpMessageTransport extends IDisposable {
 	readonly state: IObservable<McpConnectionState>;
-	readonly onDidLog: Event<string>;
+	readonly onDidLog: Event<{ level: LogLevel; message: string }>;
 	readonly onDidReceiveMessage: Event<MCP.JSONRPCMessage>;
 	send(message: MCP.JSONRPCMessage): void;
 	stop(): void;

--- a/src/vs/workbench/contrib/mcp/common/mcpServerConnection.ts
+++ b/src/vs/workbench/contrib/mcp/common/mcpServerConnection.ts
@@ -8,7 +8,7 @@ import { Disposable, DisposableStore, IReference, MutableDisposable, toDisposabl
 import { autorun, IObservable, observableValue } from '../../../../base/common/observable.js';
 import { localize } from '../../../../nls.js';
 import { IInstantiationService } from '../../../../platform/instantiation/common/instantiation.js';
-import { ILogger } from '../../../../platform/log/common/log.js';
+import { ILogger, log } from '../../../../platform/log/common/log.js';
 import { IMcpHostDelegate, IMcpMessageTransport } from './mcpRegistryTypes.js';
 import { McpServerRequestHandler } from './mcpServerRequestHandler.js';
 import { IMcpServerConnection, McpCollectionDefinition, McpConnectionState, McpServerDefinition, McpServerLaunch } from './mcpTypes.js';
@@ -66,8 +66,8 @@ export class McpServerConnection extends Disposable implements IMcpServerConnect
 
 		store.add(toDisposable(() => cts.dispose(true)));
 		store.add(launch);
-		store.add(launch.onDidLog(msg => {
-			this._logger.info(msg);
+		store.add(launch.onDidLog(({ level, message }) => {
+			log(this._logger, level, message);
 		}));
 
 		let didStart = false;

--- a/src/vs/workbench/contrib/mcp/common/mcpServerRequestHandler.ts
+++ b/src/vs/workbench/contrib/mcp/common/mcpServerRequestHandler.ts
@@ -4,13 +4,13 @@
  *--------------------------------------------------------------------------------------------*/
 
 import { equals } from '../../../../base/common/arrays.js';
-import { DeferredPromise } from '../../../../base/common/async.js';
+import { DeferredPromise, IntervalTimer } from '../../../../base/common/async.js';
 import { CancellationToken } from '../../../../base/common/cancellation.js';
 import { Emitter } from '../../../../base/common/event.js';
-import { Disposable } from '../../../../base/common/lifecycle.js';
+import { Disposable, DisposableStore } from '../../../../base/common/lifecycle.js';
 import { autorun } from '../../../../base/common/observable.js';
 import { IInstantiationService } from '../../../../platform/instantiation/common/instantiation.js';
-import { ILogger } from '../../../../platform/log/common/log.js';
+import { canLog, ILogger, LogLevel } from '../../../../platform/log/common/log.js';
 import { IProductService } from '../../../../platform/product/common/productService.js';
 import { IMcpMessageTransport } from './mcpRegistryTypes.js';
 import { McpConnectionState, MpcResponseError } from './mcpTypes.js';
@@ -81,7 +81,13 @@ export class McpServerRequestHandler extends Disposable {
 	 */
 	public static async create(instaService: IInstantiationService, launch: IMcpMessageTransport, logger: ILogger, token?: CancellationToken) {
 		const mcp = new McpServerRequestHandler(launch, logger);
+		const store = new DisposableStore();
 		try {
+			const timer = store.add(new IntervalTimer());
+			timer.cancelAndSet(() => {
+				logger.info('Waiting for server to respond to `initialize` request...');
+			}, 5000);
+
 			await instaService.invokeFunction(async accessor => {
 				const productService = accessor.get(IProductService);
 				const initialized = await mcp.sendRequest<MCP.InitializeRequest, MCP.InitializeResult>({
@@ -109,6 +115,8 @@ export class McpServerRequestHandler extends Disposable {
 		} catch (e) {
 			mcp.dispose();
 			throw e;
+		} finally {
+			store.dispose();
 		}
 	}
 
@@ -163,13 +171,21 @@ export class McpServerRequestHandler extends Disposable {
 		});
 
 		// Send the request
-		this.launch.send(jsonRpcRequest);
+		this.send(jsonRpcRequest);
 		const ret = promise.p.finally(() => {
 			cancelListener.dispose();
 			this._pendingRequests.delete(id);
 		});
 
 		return ret as Promise<R>;
+	}
+
+	private send(mcp: MCP.JSONRPCMessage) {
+		if (canLog(this.logger.getLevel(), LogLevel.Debug)) { // avoid building the string if we don't need to
+			this.logger.debug(`[editor -> server] ${JSON.stringify(mcp)}`);
+		}
+
+		this.launch.send(mcp);
 	}
 
 	/**
@@ -200,13 +216,17 @@ export class McpServerRequestHandler extends Disposable {
 	}
 
 	private sendNotification<N extends MCP.ClientNotification>(notification: N): void {
-		this.launch.send({ ...notification, jsonrpc: MCP.JSONRPC_VERSION });
+		this.send({ ...notification, jsonrpc: MCP.JSONRPC_VERSION });
 	}
 
 	/**
 	 * Handle incoming messages from the server
 	 */
 	private handleMessage(message: MCP.JSONRPCMessage): void {
+		if (canLog(this.logger.getLevel(), LogLevel.Debug)) { // avoid building the string if we don't need to
+			this.logger.debug(`[server <- editor] ${JSON.stringify(message)}`);
+		}
+
 		// Handle responses to our requests
 		if ('id' in message) {
 			if ('result' in message) {
@@ -268,7 +288,7 @@ export class McpServerRequestHandler extends Disposable {
 						message: `Method not found: ${request.method}`
 					}
 				};
-				this.launch.send(errorResponse);
+				this.send(errorResponse);
 				break;
 			}
 		}
@@ -347,7 +367,7 @@ export class McpServerRequestHandler extends Disposable {
 			id: request.id,
 			result
 		};
-		this.launch.send(response);
+		this.send(response);
 	}
 
 	/**

--- a/src/vs/workbench/contrib/mcp/test/common/mcpRegistryTypes.ts
+++ b/src/vs/workbench/contrib/mcp/test/common/mcpRegistryTypes.ts
@@ -6,6 +6,7 @@
 import { Emitter } from '../../../../../base/common/event.js';
 import { Disposable } from '../../../../../base/common/lifecycle.js';
 import { observableValue } from '../../../../../base/common/observable.js';
+import { LogLevel } from '../../../../../platform/log/common/log.js';
 import { IMcpMessageTransport } from '../../common/mcpRegistryTypes.js';
 import { McpConnectionState } from '../../common/mcpTypes.js';
 import { MCP } from '../../common/modelContextProtocol.js';
@@ -15,7 +16,7 @@ import { MCP } from '../../common/modelContextProtocol.js';
  * Allows tests to easily send/receive messages and control the connection state.
  */
 export class TestMcpMessageTransport extends Disposable implements IMcpMessageTransport {
-	private readonly _onDidLog = this._register(new Emitter<string>());
+	private readonly _onDidLog = this._register(new Emitter<{ level: LogLevel; message: string }>());
 	public readonly onDidLog = this._onDidLog.event;
 
 	private readonly _onDidReceiveMessage = this._register(new Emitter<MCP.JSONRPCMessage>());
@@ -81,7 +82,7 @@ export class TestMcpMessageTransport extends Disposable implements IMcpMessageTr
 	 * Simulate a log event.
 	 */
 	public simulateLog(message: string): void {
-		this._onDidLog.fire(message);
+		this._onDidLog.fire({ level: LogLevel.Info, message });
 	}
 
 	/**

